query-pagination-next and previous: Changing the markup on the first and last pages of a pagination

matiasbenedetto opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->
Hi :wave:

## Description
<!-- Please describe what you have changed or added -->
This PR changes the rendering of `core/query-pagination-next` and `core/query-pagination-previous` for 2 "edge cases". These cases are when the user is navigating the **first or last page** of a paginated query. In these cases, we render a `<span>` instead of nothing as we commonly do with the pagination numbers where the number matching the pagination's current number renders a `<span>` instead of an `<a>`.

After the changes featured in this PR the expected behaviour is:
- While navigating a first page of a pagination:
    - `core/query-pagination-previous` renders a `<span>` element instead of nothing.
    - `core/query-pagination-next` renders an `<a>` element as always.
    
- While navigating a last page of a pagination:
    - `core/query-pagination-previous` renders an `<a>` element as always.
    - `core/query-pagination-next` renders a `<span>` element instead of nothing.
    
- While navigating any page of pagination that is not the first or the last one:
    - `core/query-pagination-previous` renders an `<a>` element as always.
    - `core/query-pagination-next` renders an `<a>` element as always.
    
-  While navigating a page with a `core/query-pagination` block in place but no data to paginate:
    - Nothing is rendered as always.

This is just a continuation of a [work started](https://github.com/WordPress/gutenberg/pull/35430) by @MaggieCabrera

### Why are we introducing this change?
The intention is to fix the problems around the use of  `space-around` justification in the flex container of the `core/query-pagination` block as described [in this issue (#34997)](https://github.com/WordPress/gutenberg/issues/34997).



## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Tested with normal and inherit wp queries using Empty, TT1, and Blockbase themes.

Block code:
```
<!-- wp:query-pagination {"layout":{"type":"flex","justifyContent":"space-between"}} -->
	<!-- wp:query-pagination-previous /-->
	<!-- wp:query-pagination-numbers /-->
	<!-- wp:query-pagination-next /-->
<!-- /wp:query-pagination -->
```

## Screenshots <!-- if applicable -->

- First page (Empty theme):

| Before      | After |
| ----------- | ----------- |
| <img width="450" alt="image" src="https://user-images.githubusercontent.com/1310626/142674854-0396f781-bed2-492c-a0c3-021215c156ca.png">  | <img width="455" alt="image" src="https://user-images.githubusercontent.com/1310626/142674191-28e2203c-a5de-415d-b54c-213d737ab6aa.png">     |



- Last page (Empty theme):

| Before      | After |
| ----------- | ----------- |
| <img width="451" alt="image" src="https://user-images.githubusercontent.com/1310626/142674765-ba4b5d9f-c877-41a1-9e8e-c219d36875ff.png">  | <img width="460" alt="image" src="https://user-images.githubusercontent.com/1310626/142674484-a800c249-4b3d-4227-a622-55d78564bf74.png"> |




- First page (TT1 theme):

| Before      | After |
| ----------- | ----------- |
| <img width="329" alt="image" src="https://user-images.githubusercontent.com/1310626/142677319-974e9046-96fe-4978-8617-21f6e0e2e79b.png"> | <img width="324" alt="image" src="https://user-images.githubusercontent.com/1310626/142676911-27ddcfae-8ca4-48c5-840d-83f87b43afb3.png">    |






- Last page (TT1 theme):

| Before      | After |
| ----------- | ----------- |
| <img width="323" alt="image" src="https://user-images.githubusercontent.com/1310626/142677161-4f1e2219-d0af-4192-9c5f-4ee91889a4f3.png">  | <img width="318" alt="image" src="https://user-images.githubusercontent.com/1310626/142676990-30d2d651-2002-450a-8dbc-22a3d4bf8f22.png"> |




## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

MaggieCabrera committed:
add an empty wrapper when at first/last page

MaggieCabrera committed:
keep label on placeholder, add extra class

matiasbenedetto committed:
Query pagination previous link: Avoids printing the link if there are no pages to paginate

matiasbenedetto committed:
Query pagination next: removing custom 'last-page' class.

We can target the same element and know when we are navigating the last page with a selector like this: "span.wp-block-query-pagination-next"

matiasbenedetto committed:
php code style format

scruffian commented:
This is working well for me but it would be good to have some designers eyes on it - I'm not sure about the Previous/Next Page text appearing when it doesn't do anything. cc @jasmussen @kjellr

scruffian commented on packages/block-library/src/query-pagination-next/index.php:
```diff
@@ -26,7 +27,13 @@ function render_block_core_query_pagination_next( $attributes, $content, $block
 	}
-	$content = '';
+
+	$wrapper_attributes = get_block_wrapper_attributes();
```
This is already defined on line 23, do we need to call it again?

matiasbenedetto replied:
[Removed!](https://github.com/WordPress/gutenberg/pull/36681/commits/b9ab20a95cdf7f9e9529e7d39f546729f7390e5b)

scruffian commented on packages/block-library/src/query-pagination-next/index.php:
```diff
@@ -26,7 +27,13 @@ function render_block_core_query_pagination_next( $attributes, $content, $block
-	$content = '';
+
+	$wrapper_attributes = get_block_wrapper_attributes();
+	$content            = sprintf(
```
Should we escape the output here incase the user tries to insert something untrustworthy in the attribute? (https://developer.wordpress.org/themes/theme-security/data-sanitization-escaping/)

scruffian replied:
Given that its not escaped elsewhere in this file it's probably safe...

matiasbenedetto replied:
Yep, I think so

scruffian commented on packages/block-library/src/query-pagination-previous/index.php:
```diff
@@ -25,23 +27,39 @@ function render_block_core_query_pagination_previous( $attributes, $content, $bl
-	} elseif ( 1 !== $page ) {
-		$content = sprintf(
+	} else {
+		$content = 1 !== $page ? sprintf(
```
I think this goes against the brace style in the PHP guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/

matiasbenedetto replied:
fixed :)

jasmussen commented:
>  I'm not sure about the Previous/Next Page text appearing when it doesn't do anything

I would agree, the text makes it sound actionable when it isn't. I do recall Kjell having some trouble with aligning the three harmoniously in a container, and I imagine there might be a layout challenge if the 3rd one is removed entirely.

MaggieCabrera commented:
> > I'm not sure about the Previous/Next Page text appearing when it doesn't do anything
> 
> I would agree, the text makes it sound actionable when it isn't. I do recall Kjell having some trouble with aligning the three harmoniously in a container, and I imagine there might be a layout challenge if the 3rd one is removed entirely.

I think it would make sense to have it with lowered opacity. Right now it's already a challenge to keep the 3 items in the same place with any given layout if one of them disappears completely on first/last page too. And with the class added here, theme developers have the chance to style those cases in their own way too.

kjellr commented:
Yeah, having the text present but unclickable seems incorrect here, even if it's a lower opacity. The standard behavior around the internet is basically what we have already today: the text only shows up if it's possible to paginate in that direction. 

The issue we're trying to solve for here is just that we can't automatically center the page numbers if there are two items in this container. That's just not the way that `space-around` works. If these were standard blocks and I always wanted the page numbers centered, I might try using a columns block. Have we tried something like that here?

MaggieCabrera commented:
> The issue we're trying to solve for here is just that we can't automatically center the page numbers if there are two items in this container. That's just not the way that `space-around` works. If these were standard blocks and I always wanted the page numbers centered, I might try using a columns block. Have we tried something like that here?

The problem is that inside the pagination block only the query previous/next/numbers blocks are allowed, so you can't wrap them inside column blocks.

kjellr commented:
Right, but that's theoretically something we could change if it is necessary.

scruffian commented:
What if we use CSS grid?

kjellr commented:
Yeah, this seems like an appropriate use case for `grid`. Not sure the feasibility of adding that in here though. @jasmussen are you up to date on the explorations around Grid in the editor?

jasmussen commented:
Grid in terms of theme.json integration with corresponding UI is a bit backburnered and not something to consider. However it looks like the "Query Pagination" block (`wp-block-query-pagination`) comes with its own style.scss file which we could add some grid rules to. If those grid rules are sufficiently generic and utilitarian, that seems fine?

matiasbenedetto commented:
I think using `display:grid` is not the best option here. Query pagination has flex controls that work fine and are useful to customize the block rendering.
<img width="687" alt="image" src="https://user-images.githubusercontent.com/1310626/143042617-e608929d-b007-4b61-a41d-37baf25bd2fa.png">

What do you think about adding `visibility:hidden` and `aria-hidden="true"` to the non clickeable previous and next elements. I think that solution covers the three issue discussed here:
- Solves the positioning issue
- Invisible to the eye
- Invisible for the screen readers

What do you think?

matiasbenedetto committed:
removing repeated line of code

matiasbenedetto committed:
php code style

matiasbenedetto committed:
Query pagination previous and next: non clickeable elements are hidden

When we render non clickeable elements <span> instead of <a>  for the next and previous elements we make them not visible and hidden from screen readers

matiasbenedetto commented:
I implemented the change suggested in my last comment with [this commit](https://github.com/WordPress/gutenberg/pull/36681/commits/22eb053fcb01e268d877833ab191b5cc419d68ff):

Results:
- First page:
<img width="644" alt="image" src="https://user-images.githubusercontent.com/1310626/143048763-179fb34b-f235-41fc-b1e4-b64cd54c3045.png">

- Last page:
<img width="633" alt="image" src="https://user-images.githubusercontent.com/1310626/143048840-5f7f8cbb-108b-4bee-8027-e26b31996b62.png">

scruffian commented:
I like this solution. It would be good to get some input from someone from an accessibility perspective, although I'm not sure who to ask...

matiasbenedetto committed:
php lint

MaggieCabrera commented:
What happens with this solution when you align the items to the left or to the right instead of using space between?

matiasbenedetto commented:
This last case mentioned by @MaggieCabrera (the use of `justify:flex-start`), can be problematic on the first page. Because the element, even though is not visible (`visibility:hidden`), is still there. The same happens if we use `justify:flex-end` and navigate the last page.


<img width="1056" alt="image" src="https://user-images.githubusercontent.com/1310626/143077670-43918c13-1672-4202-a808-030ac4265c76.png">

scruffian commented:
> I think using display:grid is not the best option here. Query pagination has flex controls that work fine and are useful to customize the block rendering.

We could conisder mapping the flex controls to CSS grid styles without making the user aware that's what's happening...

matiasbenedetto committed:
Adding a class to the wrapper layout element rendered by Gutenberg

Adding a class describing the justify value of the layout element to be able to target the different justifications with CSS selectors.

matiasbenedetto committed:
Query pagination: hiding non visible elements when the justification is right or left

In this way we will avoid unwanted spaces when the justification is right or left.

matiasbenedetto commented:
I think that the latest 2 commits fix the remaining problem around left and right `Flex` justifications, so I think it covers all the issues discussed so far:

- Solves the positioning issue with space-around
- Invisible to the eye
- Invisible for the screen readers
- Works with left and right justified elements


Justify left  on first page:
<img width="623" alt="image" src="https://user-images.githubusercontent.com/1310626/143267184-c997ddb7-3816-4924-81a6-8a8fc842f73d.png">


Justify left on middle page:
<img width="636" alt="image" src="https://user-images.githubusercontent.com/1310626/143267291-ae308240-05fa-4f3a-a3f9-ac39dd50b82f.png">

Justify left on last page:
<img width="645" alt="image" src="https://user-images.githubusercontent.com/1310626/143267407-3d46a18d-051e-4513-903a-59dcab4fab52.png">

Space around on first page:
<img width="644" alt="image" src="https://user-images.githubusercontent.com/1310626/143267602-463b32c1-af84-4f1e-ba86-750c5420a3a7.png">

Space around on middle page:
<img width="647" alt="image" src="https://user-images.githubusercontent.com/1310626/143267712-b78f0d8a-ec92-400d-a958-bcae4e543857.png">



Space around on last page:
<img width="653" alt="image" src="https://user-images.githubusercontent.com/1310626/143267771-2efd386d-3df7-451b-bd0f-5ac5a884ccc6.png">

matiasbenedetto committed:
avoid printing class if is a value is missing

scruffian commented:
When my site has only 2 posts I see a disabled "Next Page" link on the index page:
<img width="976" alt="Screenshot 2021-11-26 at 12 45 12" src="https://user-images.githubusercontent.com/275961/143583030-a454496c-4ab0-4b0b-97ef-9a2c7fd1b5d3.png">

matiasbenedetto committed:
Fixing elements rendered on a few edge cases

matiasbenedetto committed:
removing not needed CSS

matiasbenedetto committed:
fixing css selectors

matiasbenedetto committed:
removing debugging line

matiasbenedetto commented:
The latest commits are fixing some failing use cases, as the last one spotted by @scruffian.
The block should render as the examples shared in [my previous comment](https://github.com/WordPress/gutenberg/pull/36681#issuecomment-977982001).

scruffian commented:
I no longer see the empty span on the first and last pages...

scruffian commented:
> I no longer see the empty span on the first and last pages...

This was an environment issue!

scruffian reviewed (approved):
This is looking good to me. There's some duplication between the blocks that would be good to clear up in a followup PR.

ntsekouras commented on lib/block-supports/layout.php:
```diff
@@ -160,13 +160,15 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
+	$id              = uniqid();
+	$style           = gutenberg_get_layout_style( ".wp-container-$id", $used_layout, $has_block_gap_support );
+	$container_class = 'wp-container-' . $id . ' ';
+	$justify_class   = $used_layout['justifyContent'] ? 'wp-justify-' . $used_layout['justifyContent'] . ' ' : '';
```
Why we added this to the `layout` abstraction? Was something specific to pagination blocks?

matiasbenedetto replied:
Yep, we needed to identify with a class the justification of the block, without this change that was not possible.

ntsekouras replied:
There shouldn't be any block logic leaked into the `layout` abstraction. I'll check a bit the PR better to try to understand. That was the case recently with `Navigation` block as well, but we should find another way to handle specific block needs.

matiasbenedetto replied:
Ok, I see, let me submit a PR with this change you propose.

matiasbenedetto replied:
@ntsekouras a PR moving the functionality from Layout to Query Pagination code:
https://github.com/WordPress/gutenberg/pull/37113

ntsekouras commented on packages/block-library/src/query-pagination/style.scss:
```diff
@@ -32,7 +32,17 @@ $pagination-margin: 0.5em;
 	}
 
-	&.aligncenter {
-		justify-content: center;
```
This shouldn't be removed. It was added here: https://github.com/WordPress/gutenberg/pull/34739 as a fix for `align center`.

matiasbenedetto replied:
That behavior is achieved using the centered justification:
![image](https://user-images.githubusercontent.com/1310626/144631371-2a221e9c-3e7c-4e5b-b1f3-8c03ce9dda02.png)

ntsekouras replied:
I know, but we also support the `center` alignment and is needed for backwards compatibility.

matiasbenedetto replied:
@ntsekouras could you provide more context about backward compatibility related to this, please?

ntsekouras replied:
Sure. Even if the `align:center` achieves the same think with centered justification, we still support this alignment. So it cannot be removed because it has created a regression. You can check this on trunk by selecting `align:center` and checking the front-end.

Even if we removed the `align:center` option though and with a new block migration in place, we would still need to keep the css rule because the migrations run and update the blocks markup in the editor and then we have to also save to get the new markup in the post content. So for content with blocks that had `align:center` but the migration was never applied, the content should look as before in the front end without and broken styles.

Makes sense? Please tell me to clarify anything  from the above if you want.

ntsekouras commented on packages/block-library/src/query-pagination-previous/index.php:
```diff
@@ -17,31 +17,63 @@
-		);
+	} elseif ( ! $max_page || $max_page > $page ) {
+		$custom_query  = new WP_Query( build_query_vars_from_query_block( $block, $page ) );
+		$max_num_pages = $custom_query->max_num_pages ? $custom_query->max_num_pages : 1;
```
Why do we need this `max_num_pages` logic? Previously it would enter this `if` when it was a `custom` query and it wasn't on the first page. So wouldn't a check about `$page === 1` be enough? Am I missing something?

matiasbenedetto replied:
This is necessary because if the query has no results  `$custom_query->max_num_pages` **equals to zero**. So the logic to know if there are multiple pages `1 < $max_num_pages` doesn't work without this.

ntsekouras replied:
I'm sorry but I still don't get it. In `Query Pagination Previous` we only care about whether we are on the first page or not, no? If we are on the first page, we don't have a link to show, but if we are not we always have a link. This is different from `Query Pagination Next` where we have to recreate the query run from `Query Loop` to know if we have a link for the next posts.

matiasbenedetto replied:
Nope, we don't care only about if we are on the first page or not. We need to know if there are pages to paginate. If there are not pages to paginate should not render anything.

ntsekouras commented:
Aside my comments that I'm trying to understand the changes here, I'd really want to thank you @matiasbenedetto for your contribution and effort here!  Really appreciate it that you have also opened a follow up PR 💯

ntsekouras commented:
I've created a [follow up PR](https://github.com/WordPress/gutenberg/pull/37125) to discuss/check how to simplify things (now is WIP).

ntsekouras commented:
👋 I noticed this PR when this notice was shown:

<img width="1071" alt="Screen Shot 2021-12-06 at 9 15 37 AM" src="https://user-images.githubusercontent.com/272444/144811052-b254dd2a-1c6a-49c8-b8bd-ea078ce4af4e.png">

but it made me look into the changes here. I tried to understand the added complexity, especially in the case of `inherit the global` query case, where the added logic seemed to me like trying to duplicate the [get_next_posts_link](https://developer.wordpress.org/reference/functions/get_next_posts_link/) and [get_previous_posts_link](https://developer.wordpress.org/reference/functions/get_previous_posts_link/). That led to a draft PR which [I just closed](https://github.com/WordPress/gutenberg/pull/37125#issuecomment-987046094).

Personally the concept of adding an extra element (hidden wrapper) for design purposes seems a bit odd to me and I'm not really sure if this should be added in the first place. Would another kind of layout be a better fit for these blocks (even if it doesn't exist yet like `grid`)? If yes, could this be something that a theme could accomplish with css overrides until it's properly implemented in core/GB?

Additionally it seems the `vertical` layout still renders the `placeholders` and that can't be desired I think.

https://user-images.githubusercontent.com/16275880/144903491-ed54ae3a-00af-4205-8295-12c302a173f2.mov

To sum up, I think we should consider reverting this PR and see for alternative approaches that should properly fix the design issues we want. I'm worried that if we follow this path, we just hack our way around the root issue and add complexity in block logic and styles to accommodate cases like the `flex-start` where we now `display:none`, the `vertical` layout etc..

The root issue for me is that using `flex` layout for `Query Pagination` might not just be the right layout for this block.

--cc
@youknowriad @mcsf @gziolo @ockham

matiasbenedetto commented:
Personally, I don't see how using CSS grid could make this better for users and theme devs.

I haven't tested the vertical flex orientation in this component before. The fix for the vertical orientation requires a really tiny CSS change [proposed in this PR](https://github.com/WordPress/gutenberg/pull/37163).

kjellr commented:
> To sum up, I think we should consider reverting this PR and see for alternative approaches that should properly fix the design issues we want. I'm worried that if we follow this path, we just hack our way around the root issue and add complexity in block logic and styles to accommodate cases like the flex-start where we now display:none, the vertical layout etc..
> 
> The root issue for me is that using flex layout for Query Pagination might not just be the right layout for this block.

I haven't been keeping up to date on developments since this comment, but for what it's worth, I think reverting this would be fine for now. Two thoughts: 

- The solution here is a bit complicated, and I think we can probably come up with something more ideal with a bit more time. We're leaning on the flex layouts more and more as we continue to build themes, so we may stumble on something better soon. 
- The design issue this is solving is _kind of_ awkward, but not a deal-breaker for me. The end state we reached is also a little awkward, since the homepage will still have empty space on one side. 
- Plus, in the immediate future, I think we can work around the design/layout issue by just removing the numbered pagination from the default display of the block. I don't think it's a must-have block by default, and folks can add it in manually if they require it.

mcsf commented:
> * The solution here is a bit complicated, and I think we can probably come up with something more ideal with a bit more time. We're leaning on the flex layouts more and more as we continue to build themes, so we may stumble on something better soon.
> * The design issue this is solving is _kind of_ awkward, but not a deal-breaker for me. The end state we reached is also a little awkward, since the homepage will still have empty space on one side.
> * Plus, in the immediate future, I think we can work around the design/layout issue by just removing the numbered pagination from the default display of the block. I don't think it's a must-have block by default, and folks can add it in manually if they require it.

Thanks for the design and product perspective! They match my assessment of the situation and give me confidence to also recommend reverting. Especially since we're dealing with a non-must-have issue that can be handled downstream. /cc @matiasbenedetto @jeffikus

scruffian commented:
It looks like the revert button isn't going to work on this one.
